### PR TITLE
PyImportSort: Option to treat imports separately

### PR DESCRIPTION
**Pull request by meetmangukiya**
If this option is set, then the imports separated by a line will be
fed independently to the SortImports function and the results will be
yielded accordingly.

Fixes https://github.com/coala/coala-bears/issues/879

### CODE REVIEW

File: bears/python/PyImportSortBear.py [hunk 1]
```diff
@@ -176,13 +177,36 @@ def run(self, filename, file,
...
+                                                            for x in units]),
+                                           **isort_settings)
+                new_file = tuple(sort_imports.output.splitlines(True))
+
+                if new_file != tuple(file):
+                    diff = Diff.from_string_arrays(file, new_file)
+                    yield Result(self,
+                                 'Imports can be sorted.',
+                                 affected_code=diff.affected_code(filename),
+                                 diffs={filename: diff})
```
Comment (meetmangukiya) [inline bears/python/PyImportSortBear.py]: Would be better if I could not duplicate the code, wrap the reused code in a function maybe :confused: :question:

**Comment by nemaniarjun** [comment]
@meetmangukiya Coverage didnt reach a 100%, Marking WIP :)

File: bears/python/PyImportSortBear.py [hunk 2]
```diff
@@ -176,13 +177,53 @@ def run(self, filename, file,
...
 
-        sort_imports = SortImports(file_contents=''.join(file),
-                                   **isort_settings)
-        new_file = tuple(sort_imports.output.splitlines(True))
+        if treat_seperated_imports_independently:
+            import_stmts = []
+            tmp = []
+            for lineno, lines in enumerate(file, start=1):
+                print(lines)
+                if 'import' in lines:
```
Comment (meetmangukiya) [inline bears/python/PyImportSortBear.py]: ThIs looks like a fair point

**Comment by sils** [comment]
Your commit messsage no verb.

File: bears/python/PyImportSortBear.py [hunk 3]
```diff
@@ -47,7 +49,8 @@ def run(self, filename, file,
             known_third_party_imports: typed_list(str)=(),
             known_standard_library_imports: typed_list(str)=None,
             max_line_length: int=79,
-            imports_forced_to_top: typed_list(str)=()):
+            imports_forced_to_top: typed_list(str)=(),
+            treat_seperated_imports_independently: bool=False):
```
Comment (sils) [inline bears/python/PyImportSortBear.py]: missing documentation

**Comment by meetmangukiya** [comment]
There is a problem currently..
I am splitting the import into sections by checking if given line contains `import` in it which does this:
- 
```
import this
import y

import z
``` 
gets grouped into (`import this`, `import y`) and (`import z`)

But this leads to 
- 
```
import this
import (bearlib,
            coalib)
```
 to be grouped as ('import this', 'import (bearlib, ') which is **so** wrong

**Comment by meetmangukiya** [comment]
# Reviewers 

Please see the `seperate_imports` function, that tries to solve https://github.com/coala/coala-bears/pull/1356#issuecomment-285959401 There is even a test that ensures that it works as expected, if you find any more edge cases, please report :)

File: tests/python/PyImportSortBearTest.py [hunk 4]
```diff
@@ -21,3 +29,81 @@
...
+
+    def test_import_seperation(self):
+        self.file = (
+            """
+from x import (y,
+               z,
+               w)
+import this
+
+import x, y
```
Comment (jayvdb) [inline tests/python/PyImportSortBearTest.py]: `y` here shadows `y` in the first import statement.
pyflakes would scream about that.
maybe use `a, b` instead here.

File: tests/python/PyImportSortBearTest.py [hunk 5]
```diff
@@ -21,3 +29,81 @@
...
+class PyImportSortBearTest(unittest.TestCase):
+
+    def setUp(self):
+        self.queue = Queue()
+        self.section = Section('PyImportSortBear')
+        self.uut = PyImportSortBear(self.section,
+                                    self.queue)
+
+    def test_treat_seperated_imports_independently(self):
+        self.file = (
```
Comment (jayvdb) [inline tests/python/PyImportSortBearTest.py]: no need for `self.` here, as the attribute isnt used outside of the method.

Also normally these sample files are declared as module level variables, as they can often be reused in multiple tests.

File: bears/python/PyImportSortBear.py [hunk 6]
```diff
@@ -145,6 +175,9 @@ def run(self, filename, file,
             dependencies that occur in many projects.
         :param max_line_length:
             Maximum number of characters for a line.
+        :param treat_seperated_imports_independently:
+            Treat import statements seperated by more then one line as
```
Comment (jayvdb) [inline bears/python/PyImportSortBear.py]: then -> than

"..seperated by more then one line.." -- this sounds like two blank lines needed between import groups?? that would be a pep8 violation, and isnt what is in your tests.

I think "..separated by a blank line.." is better.

File: bears/python/PyImportSortBear.py [hunk 7]
```diff
@@ -176,13 +209,38 @@ def run(self, filename, file,
             isort_settings['known_standard_library'] = (
                 known_standard_library_imports)
 
-        sort_imports = SortImports(file_contents=''.join(file),
-                                   **isort_settings)
-        new_file = tuple(sort_imports.output.splitlines(True))
+        if treat_seperated_imports_independently:
+            import_stmts = PyImportSortBear.seperate_imports(file)
```
Comment (jayvdb) [inline bears/python/PyImportSortBear.py]: the revised sorting should be a separate function, so we dont have so much logic here in the run() method.

File: bears/python/PyImportSortBear.py [hunk 8]
```diff
@@ -176,13 +209,38 @@ def run(self, filename, file,
...
+            diff = Diff(file)
+            for old, new in sorted_imps:
+                start = old[0][0]
+                end = start + len(old) - 1
+                diff.delete_lines(start, end)
+                assert isinstance(new, list)
+                diff.add_lines(start, list(new))
+
+            if diff.modified != diff._file:
+                yield Result(self,
```
Comment (jayvdb) [inline bears/python/PyImportSortBear.py]: this `Result` being yielded is identical to the existing `Result` being yielded.  The code needs to be adjusted so there is only one `Result` being yielded.

**Comment by meetmangukiya** [comment]
@jayvdb I have created another commit `Rework` since I brought very much changes to previous code, and didn't just want to throw away previous commit. If the new commit looks fine(with little modifications) then I will throw away the previous commit. I hope previous commit doesn't bother for now. :)

**Comment by meetmangukiya** [comment]
I have removed the earlier tests using verify local bear and replaced those with new ones, if that is not ok, comment.

File: bears/python/PyImportSortBear.py (hunk 6 above)
Comment (jayvdb) [inline bears/python/PyImportSortBear.py]: Not done

File: tests/python/PyImportSortBearTest.py [hunk 9]
```diff
@@ -1,23 +1,109 @@
...
+""".splitlines(True)
+
+        self.assertEqual(list(self.uut.run('', test_file)), [])
+        self.assertEqual(list(self.uut.run('',
+                                           test_file2))[0].diffs[''].modified,
+                         ['\n', 'import os\n', 'import re\n', '\n',
+                          'import requests\n'])
+
+    def test_treat_seperated_imports_independently(self):
+        test_file = (
```
Comment (jayvdb) [inline tests/python/PyImportSortBearTest.py]: It is too big. Can you create a simpler example of the bug being fixed.

**Comment by meetmangukiya** [comment]
Since there are no objections, I assume it is okay https://github.com/coala/coala-bears/pull/1356#issuecomment-287405083

File: bears/python/PyImportSortBear.py (hunk 3 above)
Comment (jayvdb) [inline bears/python/PyImportSortBear.py]: Done

File: tests/python/PyImportSortBearTest.py (hunk 9 above)
Comment (jayvdb) [inline tests/python/PyImportSortBearTest.py]: Done

File: bears/python/PyImportSortBear.py [hunk 10]
```diff
@@ -145,6 +206,9 @@ def run(self, filename, file,
             dependencies that occur in many projects.
         :param max_line_length:
             Maximum number of characters for a line.
+        :param treat_seperated_imports_independently:
+            Treat import statements seperated by one or more than one blank
```
Comment (jayvdb) [inline bears/python/PyImportSortBear.py]: This is hard to read.
'Treat' is a meaningless word.
'one or more than one' is rarely used English. 'one or more' is simpler.

Comment (meetmangukiya) [inline bears/python/PyImportSortBear.py]: I made it Treat import statements ... as independent bunch, better?

**Comment by meetmangukiya** [comment]
@jayvdb please take a look again

File: tests/python/PyImportSortBearTest.py [hunk 11]
```diff
@@ -1,29 +1,93 @@
...
-                                         ('import os\nimport sys\n',
-                                          'import os\nimport sys\n'),
-                                         ('import sys\nimport os\n',
-                                          'import sys\nimport os\n'))
-
-PyImportSortBearConfigsTest = verify_local_bear(
-    PyImportSortBear,
-    ('from os import read\nfrom sys import *\n',),
-    ('from os import read\nfrom os import *\n',),
-    settings={'combine_star_imports': True})
```
Comment (jayvdb) [inline tests/python/PyImportSortBearTest.py]: Why are a lot of test classes like `PyImportSortBearConfigsTest` being removed?
It looks like some useful test scenarios are being completely removed..?

Comment (meetmangukiya) [inline tests/python/PyImportSortBearTest.py]: None of the test cases are useful IMO.
They are just testing of the setting is being passed correctly, which is being tested in my test case and as we are moving away from verify_local_bear and since I already have written tests, I didn't want to make the test file bloated. If you want I can revert this change(I am not with this decision though).
Here, the tests are only testing the settings, and having various test cases altering each setting is not quite useful, unless we are testing the tool, which we are not, because it is not our job to test the tool.

Comment (jayvdb) [inline tests/python/PyImportSortBearTest.py]: The decision to remove the tests should be a separate issue, and definitely a separate commit.

The reason for a separate issue is I'd prefer to see all existing tests pass with your PR , so I can feel safe there is no regression hiding in here , and then we talk about killing some of the tests after.

The utility of testing the tool is a reasonable debate to have.  IMO our users need to know our settings all work correctly.  And we need to know when a setting is broken by a major release of upstream, where they may delete functionality they do not like to maintain.  Or they change the setting name.  If we are not testing these things, the breakage is seen by our users.

**Comment by jayvdb** [comment]
ack 6fa0800

**Comment by jayvdb** [comment]
@rultor merge
